### ISSUE TITLE
Serializing the Enum class object

### ISSUE THREAD

**Issue by cgrand** [EsotericSoftware/kryo#529]
Attempt to serialize the Enum class object triggers a `type cannot be null` error because of the following test `(!type.isEnum() && Enum.class.isAssignableFrom(type))` which is true for the Enum` class object. https://github.com/EsotericSoftware/kryo/blob/64d7784fac8588fcdc4b3574de77b21265c2ad77/src/com/esotericsoftware/kryo/Kryo.java#L493

**Comment by magro** [comment]
Can you provide a reproducable test case? I'm wondering how usage differs from existing enum related tests...

**Comment by cgrand** [comment]
See https://github.com/cgrand/kryo/commit/87385216e40ee3f6316a4750b6b8acf35e33a4ab

Then `mvn -Dtest=DefaultSerializersTest  test` produces:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.esotericsoftware.kryo.DefaultSerializersTest
Tests run: 23, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.482 sec <<< FAILURE! - in com.esotericsoftware.kryo.DefaultSerializersTest
testClassSerializer(com.esotericsoftware.kryo.DefaultSerializersTest)  Time elapsed: 0.006 sec  <<< ERROR!
java.lang.IllegalArgumentException: type cannot be null.
	at com.esotericsoftware.kryo.DefaultSerializersTest.testClassSerializer(DefaultSerializersTest.java:317)


Results :

Tests in error:
  DefaultSerializersTest.testClassSerializer:317 » IllegalArgument type cannot b...

Tests run: 23, Failures: 0, Errors: 1, Skipped: 0
```

**Comment by magro** [comment]
Why do you want to serialize Enum.class, in which context does this make sense?

**Comment by cgrand** [comment]
I use Kryo to migrate some (Clojure) closures between jvms. Sometimes the closed-over environment embeds some compiler state (like imports in Java and the `Enum` -> `java.lang.Enum` mapping occurs in such imports map).

**Comment by magro** [comment]
Ok, then I understand why this didn't come up earlier :-)

It doesn't seem to be that easy to fix this, because default serializers are resolved via an `isAssignableFrom` check, which means that the `Enum.class` is resolved to the registered default serializer `EnumSerializer`.

It should be possible to fix this in your app by registering the `ClassSerializer` for `Enum.class`, via `kryo.register(Enum.class, new DefaultSerializers.ClassSerializer())`.
Does this work for you?

Btw, it's recommended to set `kryo.setRegistrationRequired(true)`, so you should register all classes that shall be serialized.

**Comment by cgrand** [comment]
The set of class to serialize is open. i could add a pass to create a dictionary of classes present in a given object graph and then use this dictionary for registration and send it along the serialized payload. 
However I don't see how it could be better than using automatic registration.

**Comment by magro** [comment]
Ok, I just submitted a PR with a fix: #530. Maybe you want to have a look at it?

**Comment by cgrand** [comment]
The PR looks fine. Thanks!
